### PR TITLE
Made it possible to write body parers in Java

### LINKED ISSUE #4733: Allow request body parsing in Java
It should be possible to parse requests in Java.

### PR DESCRIPTION
**Pull request by jroper**
Fixes #4733 #1732

Refactored the Java BodyParser interface so it can be implemented by Java users to actually parse a body, rather than just wrap a Scala body parser. The parsers are injected and instantiated by Guice - this required a small amount of refactoring around anything that handles Java actions.

The Java RequestBody class was refactored such that the body is passed into it as an Object, and inferred based on type when as\* methods are called.

I didn't implement support for custom handling of multipart/form-data file parts, since that is dependent on #5037 and tracked in #5044.

### CODE REVIEW

File: documentation/manual/working/javaGuide/main/http/JavaBodyParsers.md [hunk 1]
```diff
@@ -52,6 +52,6 @@ Text based body parsers (such as **text**, **json**, **xml** or **formUrlEncoded
 
 For parsers that buffer content on disk, such as the raw parser or `multipart/form-data`, the maximum content length is specified using the `play.http.parser.maxDiskBuffer` property, it defaults to 10MB.  The `multipart/form-data` parser also enforces the text max length property for the aggregate of the data fields.
 
-You can also override the default maximum content length for a given action via the `@BodyParser.Of` annotation:
```
Comment (richdougherty) [inline documentation/manual/working/javaGuide/main/http/JavaBodyParsers.md]: Can they still set the max length with an annotation property?

Comment (jroper) [inline documentation/manual/working/javaGuide/main/http/JavaBodyParsers.md]: No.  Perhaps we could still support this if we did something like define an interface that body parser could implement called `withMaxLength` that returned a newly configured body parser with that max length.  But consider:

* The motivation for limiting the length in the body parser (as opposed to validating it in the action) is to prevent trivial DOS attacks and other misuse that could cause the application to crash.
* From a resource usage perspective, your application is only as safe as the action that buffers the largest amount of memory.
* So, generally, it doesn't make sense to increase the max buffer size at the action level, since increasing one action undermines the limits on every other action.

File: framework/src/play/src/main/java/play/mvc/BodyParsers.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+import play.http.HttpErrorHandler;
+import play.libs.F;
+import play.libs.streams.Accumulator;
+import scala.compat.java8.FutureConverters;
+
+import java.util.concurrent.CompletionStage;
+import java.util.function.Function;
+
+/**
+ * Created by jroper on 27/08/15.
```
Comment (richdougherty) [inline framework/src/play/src/main/java/play/mvc/BodyParsers.java]: This Javadoc needs updating...

File: framework/src/play/src/main/scala/play/core/Execution.scala [hunk 3]
```diff
@@ -3,7 +3,7 @@
  */
 package play.core
 
-import java.util.concurrent.ForkJoinPool
+import java.util.concurrent.{ Executor, ForkJoinPool }
```
Comment (richdougherty) [inline framework/src/play/src/main/scala/play/core/Execution.scala]: Is this change needed?

File: documentation/manual/working/javaGuide/main/http/JavaBodyParsers.md (hunk 1 above)
Comment (richdougherty) [inline documentation/manual/working/javaGuide/main/http/JavaBodyParsers.md]: I can think of examples where a maxLength would vary between actions, e.g. when different parts of a site are for different users.

One way to support maxLength annotations would be to pass the annotation object (e.g. `Optional<? extends BodyParser.Of>`) to the BodyParser when it is configured or called.

**Comment by richdougherty** [comment]
LGTM - just a couple of minor comments.

File: documentation/manual/working/javaGuide/main/http/JavaBodyParsers.md (hunk 1 above)
Comment (jroper) [inline documentation/manual/working/javaGuide/main/http/JavaBodyParsers.md]: Body parsers are applied before authenticators, so it won't matter if an action is protected by authentication or not.

File: framework/src/play/src/main/java/play/mvc/BodyParsers.java (hunk 2 above)
Comment (jroper) [inline framework/src/play/src/main/java/play/mvc/BodyParsers.java]: :white_check_mark:

File: framework/src/play/src/main/scala/play/core/Execution.scala (hunk 3 above)
Comment (jroper) [inline framework/src/play/src/main/scala/play/core/Execution.scala]: :white_check_mark:

**Comment by jroper** [comment]
Updated.

File: documentation/manual/working/javaGuide/main/http/JavaBodyParsers.md (hunk 1 above)
Comment (richdougherty) [inline documentation/manual/working/javaGuide/main/http/JavaBodyParsers.md]: Don't filters get applied before EssentialActions and EssentialActions drive the BodyParser?

Comment (jroper) [inline documentation/manual/working/javaGuide/main/http/JavaBodyParsers.md]: Filters, yes.  But authentication is typically done using action composition, which typically (and always in Java) injects the authentication code in after the body is parsed.
